homeassistant/components/fritz/common.py

```diff
@@ -224,29 +237,103 @@ def scan_devices(self, now: datetime | None = None) -> None:
+
+            if service_call.service == SERVICE_CLEANUP and mode == "stale":
+                device_hosts_list: list = await self.hass.async_add_executor_job(
+                    self.fritz_hosts.get_hosts_info
+                )
+
         except (FritzServiceError, FritzActionError) as ex:
             raise HomeAssistantError("Service or parameter unknown") from ex
         except FritzConnectionException as ex:
             raise HomeAssistantError("Service not supported") from ex
 
+        entity_reg: EntityRegistry = (
+            await self.hass.helpers.entity_registry.async_get_registry()
+        )
+
+        ha_device_list: list[
+            RegistryEntry
+        ] = self.hass.helpers.entity_registry.async_entries_for_config_entry(
+            entity_reg, config_entry.entry_id
+        )
+        entity_removed: bool = False
+
+        if mode == "stale":
+            for ha_host in ha_device_list:
+                if not _cleanup_entity_filter(ha_host):
+                    continue
+                dev_found: bool = False
+                for device_host in device_hosts_list:
+                    if ha_host.unique_id.split("_")[0] == device_host["mac"]:
+                        dev_found = True
+                        break
+                if not dev_found:
+                    _LOGGER.info(
+                        "Removing entity: %s", ha_host.name or ha_host.original_name
+                    )
+                    entity_reg.async_remove(ha_host.entity_id)
+                    entity_removed = True
+
+        elif mode == "disabled":
+            if not config_entry.pref_disable_new_entities:
+                _LOGGER.info(
+                    "System option 'Enable newly added entities' is on, nothing to do"
+                )
+                return
+
+            for ha_host in ha_device_list:
+                if ha_host.disabled and _cleanup_entity_filter(ha_host):
+                    _LOGGER.info("Removing entity: %s", ha_host.entity_id)
+                    entity_reg.async_remove(ha_host.entity_id)
+                    entity_removed = True
+
+        if entity_removed:
+            await self._remove_empty_device(entity_reg)
+
+    async def _remove_empty_device(self, entity_reg: EntityRegistry) -> None:
+        """Remove devices with no entities."""
+        device_reg: DeviceRegistry = (
+            await self.hass.helpers.device_registry.async_get_registry()
+        )
+        for device_entry in dict(device_reg.devices.items()).values():
+            if (
+                len(
+                    self.hass.helpers.entity_registry.async_entries_for_device(
+                        entity_reg,
+                        device_entry.id,
+                        include_disabled_entities=True,
+                    )
+                )
+                == 0
+            ):
+                _LOGGER.info("Removing device: %s", device_entry.name)
+                device_reg.async_remove_device(device_entry.id)
```

Comment:
This removes all empty device, doesn't it? So also empty devices from other integrations? Not sure if that is necessaryly a bad thing but... how about something like:
```python
device_entry.manufacturer == 'AVM'
```

Reply:
Optimized device removal (device search based on config_entry)